design/ozone.md

```diff
@@ -0,0 +1,258 @@
+ * The real k8s golang objects are generated
+ * After final adjustment in golang, we can post all the objects to the K8S api
+
+### CRD
+
+A good operator CRD provides a good balance between the simplicity and power. By default the CRD should be as simple as possible but we should keep the possibility to customize any segment of the generated cluster.
+
+This can be achieved by the use k8s generator tool.
+
+ 1. The default cluster will be generated based on the main attributes of CRD
+ 2. Additional attributes can trigger additional transformations
+ 3. Finally we can provide an option to define any custom transformations
+
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-ozone
+---
+apiVersion: ozone.rook.io/v1alpha1
+kind: OzoneObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ozone
+spec:
+  components: #component specific settings
+     csi:
+        enable: true
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     recon:
+        enable: true
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     s3g: 
+        enable: true
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     scm:
+        #required component, always enabled
+        placement: #see rook.io/v1alpha1 PlacementSpec
+     om:
+        #required component, always enabled
+        placement: #see rook.io/v1alpha1 PlacementSpec
+
+  features: # these are not components but confiuration groups
+     prometheus: true
+     tracing: true
+  ozoneVersion: #see ceph versioning doc for mor details
+    image: apache/ozone:0.4.1
+    allowUnsupported: false
+  upgradePolicy: #see ceph versioning doc for mor details (Won't be implmented in Phase 1)
+    ozoneVersion:
+      image: apache/ozone:0.5.0
+      allowUnsupported: false
+  storage: #see rook.io/v1alpha1/StorageScopeSpec
+  placement: #see rook.io/v1alpha1 PlacementSpec
+  annotations: #custom annotations for all the created k8s resources
+     key: value
+  customization: #deep water: any additional transformations to modify k8s resources. Only for advanced users.
+    - type: Add
+      path: ["metadata", "labels"]
+      value:
+         generated-with: rook
+    
+```
+
+The type of the `storage` attribute is `StorageScopeSpec` which is a [common rook type](https://github.com/rook/rook/blob/818ad9c7ca4d352a9bed202e619d47b05f1a4dab/pkg/apis/rook.io/v1alpha2/types.go), and can be used to specifiy the details of the storage (eg. volumeClaimTemplates, nodeCounts, etc.)
+
+## Operator capabilities
+
+The main benefit to use an operator is to provide additional functionalities based on the current state of the runtime environment. The power of the operator based on the runtime information which is not available for any other installer tool.
+
+Some option which can be implemented as part of the Operator:
+ 
+ * Use emptyDir instead of Persistent Volumes in case of the dev cluster (_opt-in_ only, it should be requested by a CRD attribute)
```

Comment:
You can consider enabling this feature via setting an environment variable in the operator instead of in the CRD to make it more hidden from end users. e.g., `TESTING_USE_EMPTY_DIR_VOLS` or something

Reply:
Thanks, good idea. I modified this line.

One limitation of the ENV approach is that you can't deploy two type of clusters (real and ephemeral) with the help of the same operator instance.

But as the main target of this feature is to make the evaluation easier, it shouldn't be a problem.